cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/token/MultiClusterReplicaAwareFailureHandler.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.cassandra.spark.bulkwriter.token;
+
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Range;
+import org.apache.commons.lang.NotImplementedException;
+
+import org.apache.cassandra.spark.bulkwriter.ClusterInfo;
+import org.apache.cassandra.spark.bulkwriter.JobInfo;
+import org.apache.cassandra.spark.bulkwriter.coordinatedwrite.CassandraClusterInfoGroup;
+import org.apache.cassandra.spark.bulkwriter.coordinatedwrite.CoordinatedWriteConf;
+import org.apache.cassandra.spark.bulkwriter.coordinatedwrite.CoordinatedWriteConf.ClusterConf;
+import org.apache.cassandra.spark.common.model.CassandraInstance;
+import org.apache.cassandra.spark.data.ReplicationFactor;
+import org.apache.cassandra.spark.data.partitioner.Partitioner;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * A ReplicaAwareFailureHandler that can handle multiple clusters, including the case of single cluster.
+ *
+ * @param <I> CassandraInstance type
+ */
+public class MultiClusterReplicaAwareFailureHandler<I extends CassandraInstance> extends ReplicaAwareFailureHandler<I>
+{
+    // default failure handler for CassandraInstance w/o specifying the belonging clusterId
+    private final SingleClusterReplicaAwareFailureHandler<I> defaultFailureHandler;
+    private final Map<String, SingleClusterReplicaAwareFailureHandler<I>> failureHandlerPerCluster = new HashMap<>();
+
+    private final Partitioner partitioner;
+
+    public MultiClusterReplicaAwareFailureHandler(Partitioner partitioner)
+    {
+        this.partitioner = partitioner;
+        this.defaultFailureHandler = new SingleClusterReplicaAwareFailureHandler<>(partitioner, null);
+    }
+
+    @Override
+    public synchronized void addFailure(Range<BigInteger> tokenRange, I instance, String errMessage)
+    {
+        if (instance.hasClusterId())
+        {
```

Comment:
NIT, since we will use the value
```suggestion
        String clusterId = instance.clusterId();
        if (clusterId != null)
        {
```